cmd/sidecars/qemu-args/main.go

```diff
@@ -0,0 +1,105 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/spf13/pflag"
+
+	vmSchema "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	domainSchema "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+const (
+	qemuCommandLineArgsAnnotation = "libvirt.vm.kubevirt.io/qemuArgs"
+	NS_SCHEMA_QEMU_1_0            = "http://libvirt.org/schemas/domain/qemu/1.0"
+)
+
+func onDefineDomain(logger *log.Logger, vmiJSON, domainXML []byte) ([]byte, error) {
+	vmiSpec := vmSchema.VirtualMachineInstance{}
+	if err := json.Unmarshal(vmiJSON, &vmiSpec); err != nil {
+		logger.Printf("Failed to unmarshal given VMI spec: %s due %s", vmiJSON, err)
+		return nil, err
+	}
+
+	// Same struct as api.Commandline but with a nicer name for JSON. User can set env's name and
+	// values, or arg's multiple values in a json format. See:
+	// https://libvirt.org/drvqemu.html#pass-through-of-arbitrary-qemu-commands
+	qemuCmd := struct {
+		QEMUEnv []api.Env `json:"env,omitempty"`
+		QEMUArg []api.Arg `json:"arg,omitempty"`
+	}{}
+
+	annotations := vmiSpec.GetAnnotations()
+	if qemuArgs, found := annotations[qemuCommandLineArgsAnnotation]; !found {
+		logger.Println("No command line arguments provided. Returning original domain spec.")
+		return domainXML, nil
+	} else if err := json.Unmarshal([]byte(qemuArgs), &qemuCmd); err != nil {
+		logger.Printf("Failed to unmarshal qemu args: %s due %s", qemuArgs, err)
+		return domainXML, err
+	}
+
+	domainSpec := domainSchema.DomainSpec{}
+	if err := xml.Unmarshal(domainXML, &domainSpec); err != nil {
+		logger.Printf("Failed to unmarshal given domain spec: %s due %s", domainXML, err)
+		return nil, err
+	}
+	if domainSpec.QEMUCmd == nil {
+		domainSpec.QEMUCmd = &domainSchema.Commandline{}
+	}
+	// Set decoded annotations as QEMU Command line options
+	*domainSpec.QEMUCmd = api.Commandline(qemuCmd)
```

Comment:
This will essentially overwrite the command line. KubeVirt may already have smth set here. Perhaps an `append` would be more appropriate, wdyt? But we need to ensure that the args/env are not re-added each time (i.e. they are added only once on the first hook call).

Reply:
> But we need to ensure that the args/env are not re-added each time 

Ref: https://github.com/kubevirt/kubevirt/blob/3f7fa708521193396f759be35a3ad93cf8a8f1d0/cmd/network-slirp-binding/domain/domain.go#L85

Reply:
>But we need to ensure that the args/env are not re-added each time

UPD: ... but after thinking a bit more and looking at the actual code, it seems that we cannot get the hook called twice on the same domain, so probably that part is not relevant. Or...?

Reply:
I agree, appending is likely what we should do here to avoid removing settings from KubeVirt or other `sidecar` that has run prior to this one, if any.